spring-kafka/src/main/java/org/springframework/kafka/listener/adapter/MessagingMessageListenerAdapter.java

```diff
@@ -558,6 +637,63 @@ private void sendReplyForMessageSource(Object result, String topic, Message<?> s
 		this.replyTemplate.send(builder.build());
 	}
 
+	protected void asyncSuccess(@Nullable Object result, String replyTopic, Message<?> source, boolean returnTypeMessage) {
+		if (result == null) {
+			if (this.logger.isDebugEnabled()) {
+				this.logger.debug("Async result is null, ignoring");
+			}
+		}
+		else {
+			sendResponse(result, replyTopic, source, returnTypeMessage);
+		}
+	}
+
+	protected void acknowledge(@Nullable Acknowledgment acknowledgment) {
+		if (acknowledgment != null) {
+			acknowledgment.acknowledge();
+		}
+	}
+
+	protected void asyncFailure(Object request, @Nullable Acknowledgment acknowledgment, Consumer<?, ?> consumer,
+			Throwable t, Message<?> source) {
+		try {
+			handleException(request, acknowledgment, consumer, source,
+					new ListenerExecutionFailedException(createMessagingErrorMessage(
+							"Async Fail", source.getPayload()), t));
+			return;
+		}
+		catch (Exception ex) {
+		}
+		this.logger.error(t, () -> "Future, Mono, or suspend function was completed with an exception for " + source);
+		acknowledge(acknowledgment);
```

Comment:
Why do we ack positively in case of error?
Spring AMQP logic is like this:
```
	protected void asyncFailure(Message request, Channel channel, Throwable t, Object source) {
		this.logger.error("Future, Mono, or suspend function was completed with an exception for " + request, t);
		try {
			channel.basicNack(request.getMessageProperties().getDeliveryTag(), false,
					ContainerUtils.shouldRequeue(this.defaultRequeueRejected, t, this.logger));
		}
		catch (IOException e) {
			this.logger.error("Failed to nack message", e);
		}
	}
```
Don't we have something like that in Spring Kafka as well?

Reply:
I'm having some trouble, function `Acknowlege.nack` not support out-of-order commit now.
Another Similar functions is recoverer(in particular DLT), but not sure how to implement it yet.

Could we temporary use KafkaListenerErrorHandler?

Reply:
OK. I see you do use it in the `handleException()`:
```
Object errorResult = this.errorHandler.handleError(message, e, consumer, acknowledgment);
```

Then it goes to the:
```
handleResult(Objects.requireNonNullElse(result, errorResult),
							records, acknowledgment, consumer, message);
```

So, probably it is too late over here to make some assumption and do this unconditional `acknowledge(acknowledgment)`.

I mean it feel like this call has to be removed from here